app/controllers/chargeback_controller.rb

```diff
@@ -273,9 +300,136 @@ def cb_rate_form_field_changed
+        @edit[:selected_rd_remove].nil? ? enable_button = false : enable_button = true
+      end
+      render :update do |page|
+        if enable_button
+          page << javascript_enable_field("btm-remove_rd")
+        else
+          page << javascript_disable_field("btm-remove_rd")
+        end
+        page << javascript_unchecked("select_all")
+        page << javascript_checked_by_class("select[value=#{params[:index].to_i}]")
+      end
+    end
+  end
+
+  # AJAX driven routine to check for changes in ANY field on the form off add rate detail in rate
+  def cb_add_rd_form_field_changed
+    @edit = session[:edit]
+    render :update do |page|
+      if !params[:metric].nil?
+        if params[:metric] != "null"
+          @edit[:selected_metrics] = params[:metric]
+        end
+        page << javascript_for_cb_button_add_metric_visibility(params[:metric] != "null")
+      elsif !params[:group].nil?
+        if @edit[:new][:level] == 'Compute' || @edit[:new][:level].nil?
+          rate_details_level = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == "Compute" }
+        else
+          rate_details_level = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == "Storage" }
+        end
+        @edit[:new][:metrics] = chargeback_details_metrics(rate_details_level, params[:group])
+        @edit[:new][:group] = params[:group]
+        page.replace_html("add_metric_fields", :partial => "cb_rate_add_metrics")
+      elsif !params[:level].nil?
+        if params[:level] == 'Compute'
+          rate_details_level = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == "Compute" }
+        else
+          rate_details_level = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == "Storage" }
+        end
+        @edit[:new][:groups] = chargeback_details_groups(rate_details_level)
+        @edit[:new][:level] = params[:level]
+        # showing all the metrics of the selected group
+        if @edit[:new][:groups].empty?
+          @edit[:new][:metrics] = []
+        else
+          @edit[:new][:metrics] = chargeback_details_metrics(rate_details_level, @edit[:new][:groups].first[1])
+        end
+        page.replace_html("add_metric_fields", :partial => "cb_rate_add_metrics")
+      end
+    end
+  end
+
+  def cb_add_rd
+    @edit = session[:edit]
+    # for save added rate details
+    @added_rds = []
+    case params[:button]
+    when "add_single"
+      selected_metrics = @edit[:selected_metrics].split(',')
+      selected_metrics.each do |sm|
+        # if the detail is not selectable by metric we select by description (for the Fixed rates datail)
+        rate_detail = @sb[:rate_details].select { |k| k.metric == sm || k.description == sm }
+        rate_detail[0].rate = 1.0
+        @added_rds.push(rate_detail[0])
+      end
+    when "add_all"
+      @added_rds = @sb[:rate_details].select { |k| k.rate == "0" }
+      @added_rds.each do |r|
+        r.rate = 1.0
```

Comment:
Same here, it seems like add_single is similar to add_all for just a single rate detail.  Maybe extracting methods would help separate the common logic from the differences.